backend/resources/org/akvo/dash/migrations_tenants/007-database-functions.up.sql

```diff
@@ -39,50 +26,169 @@ BEGIN
   RETURNS jsonb AS
-$BODY$
+$$
 BEGIN
-  RETURN to_date(val, format, false);
+  RETURN initcap(val::text);
 END;
-$BODY$
+$$
   LANGUAGE plpgsql IMMUTABLE STRICT;
 
-CREATE OR REPLACE FUNCTION titlecase(val jsonb)
+CREATE OR REPLACE FUNCTION lumen_trim_double(val jsonb)
   RETURNS jsonb AS
-$BODY$
+$$
 BEGIN
-  RETURN initcap(val::text);
+  RETURN regexp_replace(val::text, '\s+', ' ', 'g');
 END;
-$BODY$
+$$
   LANGUAGE plpgsql IMMUTABLE STRICT;
 
-CREATE OR REPLACE FUNCTION trim_space(val jsonb)
+CREATE OR REPLACE FUNCTION lumen_trim(val jsonb)
   RETURNS jsonb AS
-$BODY$
+$$
 BEGIN
-  RETURN regexp_replace(val::text, '\s+', ' ', 'g');
+  RETURN regexp_replace(val::text, '^"(\s+)|(\s+)"$', '"', 'g');
 END;
-$BODY$
+$$
   LANGUAGE plpgsql IMMUTABLE STRICT;
+
+
+CREATE OR REPLACE FUNCTION lumen_get_update_sql(tbl_name text, col_name text, new_type text, on_error text)
+  RETURNS text AS
+$$
+DECLARE
+  initial_sql text = format('UPDATE %I SET %I = lumen_to_%s', tbl_name, col_name, new_type);
+BEGIN
+  IF new_type = 'text' THEN
+    RETURN format('%s(%s) WHERE rnum = $1', inital_sql, col_name);
+  ELSIF new_type = 'number' THEN
+    IF on_error = 'default-value' THEN
+      RETURN format('%s(%s, $1, false) WHERE rnum = $2', initial_sql, col_name);
+    ELSIF on_error = 'fail' OR on_error = 'delete-row' THEN
+      RETURN format('%s(%s, $1, true) WHERE rnum = $2', initial_sql, col_name);
+    ELSE
+      RAISE EXCEPTION 'Unknown on_error strategy: %', on_error;
+    END IF;
+  ELSIF new_type = 'date' THEN
+    IF on_error = 'default-value' THEN
+      RETURN format('%s(%I, $1, $2, false) WHERE rnum = $3', initial_sql, col_name);
+    ELSEIF on_error = 'fail' OR on_error = 'delete-row' THEN
+      RETURN format('%s(%I, $1, $2, true) WHERE rnum = $3', initial_sql, col_name);
+    ELSE
+      RAISE EXCEPTION 'Unknown on_error strategy: %', on_error;
+    END IF;
+  ELSE
+    RAISE EXCEPTION 'Unknown data type: %', new_type;
+  END IF;
+END;
+$$
+  LANGUAGE plpgsql IMMUTABLE STRICT;
+
+CREATE OR REPLACE FUNCTION lumen_change_data_type(tbl_name text, col_name text,
+  new_type text, default_value jsonb, parse_format text, on_error text) RETURNS jsonb AS
+$$
+DECLARE
+  cur refcursor;
+  current_rnum integer;
+  current_val jsonb;
+  exec_log text[];
+  select_sql text;
+  delete_sql text;
+  delete_flag text = '''{"delete":true}''::jsonb';
+  flag_sql text;
+  update_sql text;
+  t1 text;
+  t2 text;
+  t3 text;
+BEGIN
+  -- tbl_name: table name on which to operate
+  -- col_name: column name to change on which to change the values
+  -- new_type: 'text' | 'number' | 'date'
+  -- on_error: 'default-value' | 'fail' | 'delete-row'
+
+  IF tbl_name IS NULL OR tbl_name = '' THEN
+    RAISE EXCEPTION 'tbl_name is required';
+  END IF;
+
+  IF col_name IS NULL OR char_length(col_name) = 0 THEN
+    RAISE EXCEPTION 'col_name is required';
+  END IF;
+
+  IF new_type IS NULL OR char_length(new_type) = 0 THEN
```

Comment:
Yes. There has been several requests to implement _assertions_ in plpgpsql, but `Assert` is already a reserved word in PG, so this checks become verbose.